Single trace line source

gsgall opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

Closes #28176

gsgall committed:
Adding support for tag vector which is not zeroed automatically

gsgall committed:
Adding a test for the non zeroed vector tag

gsgall committed:
Adding the gold file for the non-zeroed vector tag

gsgall committed:
Adding an initial version of a single trace ray study test

gsgall committed:
Added testing for contributing to residuals with a line source while only tracing the rays a single time

Closes #28176

lindsayad commented on framework/include/problems/SubProblem.h:
```diff
@@ -1080,6 +1093,9 @@ class SubProblem : public Problem
   bool _have_ad_objects;
 
+  /// the list of vector tags that the application developer wants to be responsible for zeroing
+  std::set<TagID> _not_zeroed_tagged_vectors;
```
does this need to be ordered? If not, I suggest using `unordered_set`

gsgall replied:
No there is no need for this to be an ordered set since we just need it to store ids and to check if they exist within the set. I will update this.

lindsayad commented on framework/include/problems/SubProblem.h:
```diff
@@ -144,6 +144,19 @@ class SubProblem : public Problem
+   * when other tagged vectors are
+   * @param tag the TagID of the vector that will be manually managed
+   */
+  virtual void addNotZeroedVectorTag(const TagID & tag);
```
```suggestion
  void addNotZeroedVectorTag(const TagID tag);
```

lindsayad commented on framework/include/problems/SubProblem.h:
```diff
@@ -144,6 +144,19 @@ class SubProblem : public Problem
+   * when other tagged vectors are
+   * @param tag the TagID of the vector that is currently being checked
+   */
+  bool vectorTagNotZeroed(const TagID & tag);
```
```suggestion
  bool vectorTagNotZeroed(const TagID tag);
```

lindsayad replied:
It's faster to pass builtins as values as opposed to references

lindsayad reviewed (changes requested):
Do you need all the framework changes? What if you used a solution vector tag instead of a residual? It feels weird to have a residual vector that doesn't depend on the nonlinear degrees of freedom. I think the not-zeroing feels like a code smell that I'm hoping we can avoid

gsgall commented:
> Do you need all the framework changes? What if you used a solution vector tag instead of a residual? It feels weird to have a residual vector that doesn't depend on the nonlinear degrees of freedom. I think the not-zeroing feels like a code smell that I'm hoping we can avoid


I don't believe that we can use a solution tag vector since you're not able to have the vector tag of a kernel be a solution vector tag they need to be residual vector tags. 

As for the residual vector not relying on the nonlinear degrees of freedom. Our particles are moved based on the velocity that they store in ray data and this is only update on TIMESTEP_BEGIN in our simulation the particles are then moved assuming this constant velocity during the time step. This is because we are using an explicit PIC scheme right now. If we were to implement an implicit method then we would need to retrace the rays multiple times since we would be updating them with the fields. But because the particle velocities only depend on the state of the fields at the being of the time step and are constant throughout the trace, tracing on every residual evaluation results in wasted work.

gsgall committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

gsgall committed:
Addressing comments from review

loganharbour commented:
> I think the not-zeroing feels like a code smell that I'm hoping we can avoid

The alternative is that we override when a vector zeroed and let the residual call zero it for us. I've thought about this quite a bit and unfortunately can't find a particularly pretty solution.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2338792/) on b4f440f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28177/site)

This comment will be updated on new commits.

loganharbour commented on modules/ray_tracing/test/tests/raykernels/line_source_ray_kernel/tests:
```diff
@@ -2,11 +2,21 @@
+    [single_trace_test]
+      type = 'Exodiff'
+      input = 'single_trace_line_source_ray_kernel.i'
+      exodiff = 'single_trace_line_source_ray_kernel_out.e  '
```
```suggestion
      exodiff = 'single_trace_line_source_ray_kernel_out.e'
```

gsgall committed:
Update modules/ray_tracing/test/tests/raykernels/line_source_ray_kernel/tests

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

lindsayad commented:
> This is because we are using an explicit PIC scheme right now.

If this is the case, shouldn't you only have a single Newton iteration? And if you're worried about the extra nonlinear residual evaluation at the end of the iteration, you could pass `-snes_type ksponly`?

gsgall commented:
> > This is because we are using an explicit PIC scheme right now.
> 
> If this is the case, shouldn't you only have a single Newton iteration? And if you're worried about the extra nonlinear residual evaluation at the end of the iteration, you could pass `-snes_type ksponly`?

Well we are updating the particle data with an explicit scheme, I should have been more specific on this. We update particle data using a finite difference scheme before we trace them. But there may be other terms which are not particle sources that couple into the field equations. Additionally, I've seen in some of my testing that even when we are computing the electrostatic potential from a charge density which has a contribution to the potentials residual that is constant during the course of a solve, several non linear iterations can be required.

lindsayad commented on framework/include/problems/SubProblem.h:
```diff
@@ -144,6 +144,21 @@ class SubProblem : public Problem
+   * when other tagged vectors are
+   * @param tag the TagID of the vector that will be manually managed
+   */
+  void addNotZeroedVectorTag(const TagID & tag);
```
```suggestion
  void addNotZeroedVectorTag(const TagID tag);

```

lindsayad commented on framework/include/problems/SubProblem.h:
```diff
@@ -1080,6 +1095,9 @@ class SubProblem : public Problem
   /// AD flag indicating whether **any** AD objects have been added
   bool _have_ad_objects;
 
+  /// the list of vector tags that the will not be zeroed when all other tags are
```
```suggestion
  /// the list of vector tags that will not be zeroed when all other tags are
```

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -543,7 +548,6 @@ FEProblemBase::FEProblemBase(const InputParameters & parameters)
     _is_petsc_options_inserted = true;
   }
 }
-
```
Please keep this newline

gsgall replied:
Added this newline back

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -255,6 +255,11 @@ FEProblemBase::validParams()
+  params.addParam<std::vector<std::vector<TagName>>>(
+      "not_zeroed_tag_vectors",
+      {},
+      "Extra vector tags which the sytem will not zero when other vector tags are zeroed");
```
I would mention what the outer index is for (nonlinear system)

gsgall replied:
Added the same additional description as for extra_tag_vectors.

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -565,6 +569,15 @@ FEProblemBase::createTagVectors()
+    for (auto & vector : not_zeroed_vectors[nl_sys_num])
+    {
+      auto tag = addVectorTag(vector);
+      _nl[nl_sys_num]->addVector(tag, false, GHOSTED);
```
```suggestion
      _nl[nl_sys_num]->addVector(tag, false, PARALLEL);
```
If you're feeling brave I would try changing this on line 569 as well

gsgall replied:
If you wouldn't mind explaining, for my own knowledge, what is the difference between these two settings?

lindsayad replied:
It's annoying that github hid this conversation even though it wasn't resolved. If a vector is ghosted, then you can read entries in the vector that are not owned by your processor. It would have been nice if I had told you this before you ran into your issues below!!

lindsayad replied:
As noted in https://github.com/idaholab/moose/pull/28177#issuecomment-2303804746 let's revert my suggestion. My apologies!

lindsayad commented on framework/src/problems/SubProblem.C:
```diff
@@ -135,6 +135,20 @@ SubProblem::vectorTagExists(const TagName & tag_name) const
 }
 
+void
+SubProblem::addNotZeroedVectorTag(const TagID & tag)
```
```suggestion
SubProblem::addNotZeroedVectorTag(const TagID tag)
```

lindsayad commented on framework/src/problems/SubProblem.C:
```diff
@@ -135,6 +135,20 @@ SubProblem::vectorTagExists(const TagName & tag_name) const
+}
+
+bool
+SubProblem::vectorTagNotZeroed(const TagID & tag)
```
```suggestion
SubProblem::vectorTagNotZeroed(const TagID tag)
```

lindsayad commented on framework/src/problems/SubProblem.C:
```diff
@@ -135,6 +135,20 @@ SubProblem::vectorTagExists(const TagName & tag_name) const
+bool
+SubProblem::vectorTagNotZeroed(const TagID & tag)
+{
+  if (_not_zeroed_tagged_vectors.size() == 0)
+    return false;
+  return _not_zeroed_tagged_vectors.find(tag) != _not_zeroed_tagged_vectors.end();
```
```suggestion
  return _not_zeroed_tagged_vectors.count(tag);
```

lindsayad commented on modules/ray_tracing/src/userobjects/RayTracingStudy.C:
```diff
@@ -92,6 +92,10 @@ RayTracingStudy::validParams()
 
+  params.addParam<bool>("allow_other_flags_with_prekernels",
+                        false,
+                        "Whether or not to error when a user provides a list of execition flags "
```
```suggestion
                        "Whether or not to error when a user provides a list of execution flags "
```

lindsayad commented on modules/ray_tracing/src/userobjects/RayTracingStudy.C:
```diff
@@ -92,6 +92,10 @@ RayTracingStudy::validParams()
                         "Trace intersections are not verified regardless of this parameter in "
                         "optimized modes (opt, oprof).");
 
+  params.addParam<bool>("allow_other_flags_with_prekernels",
+                        false,
+                        "Whether or not to error when a user provides a list of execition flags "
+                        "and PRE_KERNELS is in the list of flags.");
```
Based on the documentation string, I would expect the parameter name to be something like `"allow_prekernels_in_exec_flags"`

gsgall replied:
I think my documentation string may be unclear. Currently the raytracing study will error if you provide the PRE_KERNELS exec flag in a list that includes other flags as well. It is totally fine to have PRE_KERNELS in the exec flags so long as it is the only exec flag.

gsgall replied:
I updated the doc string to make this more explicitly clear

lindsayad commented on modules/ray_tracing/test/src/userobjects/SingleTraceLineSourceTest.C:
```diff
@@ -0,0 +1,50 @@
+    return;
+  }
+
+  if (_fe_problem.currentlyComputingJacobian())
```
Maybe just exclude `EXEC_NONLINEAR` from the execution flags?

gsgall replied:
It looks like this works. But would you be able to tell me about what the difference between these two implementations is?

lindsayad replied:
Essentially my suggestion doesn't create the condition that you need to fix

lindsayad commented on modules/ray_tracing/test/tests/raykernels/line_source_ray_kernel/tests:
```diff
@@ -2,11 +2,21 @@
+      type = 'Exodiff'
+      input = 'line_source_ray_kernel.i'
+      exodiff = 'line_source_ray_kernel_out.e'
+      detail = 'while tracing the ray on every residual evaluation.'
```
```suggestion
      detail = 'while tracing the ray on every residual evaluation, and'
```

lindsayad commented on test/tests/tag/tests:
```diff
@@ -2,6 +2,13 @@
 
+  [not_zeroed_vector_tag]
+    type = Exodiff
+    requirement = 'The system shall support the ability for a the user to create a tagged vector which is not automatically zeroed'
```
```suggestion
    requirement = 'The system shall support the ability for a the user to create a tagged vector which is not automatically zeroed.'
```

lindsayad reviewed (commented):
Ok, I'm down with the framework changes

gsgall committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

gsgall committed:
Making some changes to address comments

- Changed the logic in the single trace line source study
- Added clarity for the doc string that allows for other flags when the execution flag for PRE_KERNELS is set

gsgall commented:
@loganharbour @lindsayad I think this should be good go now.

lindsayad reviewed (dismissed):
assuming CI passes

gsgall committed:
Updating test for clarity

- Test uses a single dirac kernel with two vector tags the variable u is a dummy variable that is not actually solved for
- Since the vector tags of the dirac kernel only has a non zero value for the first test the two tags should be the same after the first step and different on the second step

gsgall commented:
I think the framework test I had for this functionality was poorly designed. I have updated it in favor of a simpler test.

The new test uses a 2 element mesh $x\in[-1,1]$ with a single Dirac kernel at the point $x=0$ which has a value of 1 on the initial nonlinear evaluation and 0 after that, its vector tags are set to be the not zeroed tag and a normal extra vector tag. I use two time steps, the reason for this is to compare the result of the two tags. On the first nonlinear evaluation step both vector tags should have the same value of 0 on the left and right most nodes and a value of -1 at the center node. Since the Dirac kernel has a value of zero on the second nonlinear evaluation the not zeroed tag will retain the same value as it had after the first nonlinear evaluation and the regular vector tag will be reset and have a value of zero at all nodes.

gsgall commented:
@lindsayad  Is there anything that I need to do to address these failures. It looks like they are unrelated to my new feature.

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2355761/) on cf1b7ab : invalidated by @lindsayad

moosebuild commented:
Job [LibTorch](https://civet.inl.gov/job/2355758/) on cf1b7ab : invalidated by @lindsayad

lindsayad commented:
I kind of think the failures are from this PR. They are repeatable

gsgall commented:
> I kind of think the failures are from this PR. They are repeatable

I'll look into this more then.

gsgall commented:
I've done a bit more digging on where the errors in the modules are coming from with this change. It seems like the issue with trying to change the vector tag types from `GHOSTED` to `PARALLEL` 

# Solid Mechanics  
 
In this module the failing test seems to be trying access some dofs which do not exist this is the message that is produced when running this with two mpi ranks 

```
No index 18 in ghosted vector.
Vector contains [0,12)
And empty ghost array
```

and this the stack trace 

```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x90)
  * frame #0: 0x000000010858d7ec libmesh_opt.0.dylib`libMesh::PetscVector<double>::map_global_to_local_index(unsigned long long) const + 148
    frame #1: 0x0000000108588d24 libmesh_opt.0.dylib`libMesh::PetscVector<double>::get(std::__1::vector<unsigned long long, std::__1::allocator<unsigned long long>> const&, double*) const + 88
    frame #2: 0x0000000105b9dc0c libmoose-opt.0.dylib`MooseVariableScalar::reinit(bool) + 2136
    frame #3: 0x0000000105d2ec7c libmoose-opt.0.dylib`SystemBase::reinitScalars(unsigned int, bool) + 60
    frame #4: 0x0000000105710c44 libmoose-opt.0.dylib`FEProblemBase::reinitScalars(unsigned int, bool) + 284
    frame #5: 0x0000000105d53514 libmoose-opt.0.dylib`AuxiliarySystem::computeScalarVars(MooseEnumItem) + 284
    frame #6: 0x0000000105d31470 libmoose-opt.0.dylib`AuxiliarySystem::compute(MooseEnumItem) + 136
    frame #7: 0x0000000105742de4 libmoose-opt.0.dylib`FEProblemBase::computeSystems(MooseEnumItem const&) + 192
    frame #8: 0x000000010572cf38 libmoose-opt.0.dylib`FEProblemBase::execute(MooseEnumItem const&) + 308
    frame #9: 0x00000001051d6d28 libmoose-opt.0.dylib`FixedPointSolve::solveStep(double&, double&, std::__1::set<unsigned long long, std::__1::less<unsigned long long>, std::__1::allocator<unsigned long long>> const&) + 2180
    frame #10: 0x00000001051d5e1c libmoose-opt.0.dylib`FixedPointSolve::solve() + 2340
    frame #11: 0x0000000105bbb380 libmoose-opt.0.dylib`TimeStepper::step() + 44
    frame #12: 0x00000001051dbe04 libmoose-opt.0.dylib`Transient::takeStep(double) + 180
    frame #13: 0x00000001051db658 libmoose-opt.0.dylib`Transient::execute() + 140
    frame #14: 0x0000000105f1c618 libmoose-opt.0.dylib`MooseApp::executeExecutioner() + 324
    frame #15: 0x0000000105f1920c libmoose-opt.0.dylib`MooseApp::run() + 1988
    frame #16: 0x00000001029c3be4 solid_mechanics-opt`void Moose::main<SolidMechanicsTestApp>(int, char**) + 152
    frame #17: 0x00000001029c3b40 solid_mechanics-opt`main + 12
    frame #18: 0x000000018d023154 dyld`start + 2476
```
# Optimization 

I am having a hard time tracking this error down properly since I am not sure what the right assert to set in the debugger is but when running in debug mode all of the optimization test failures are failing on the same timpi assertion 

```
Assertion `this->verify(r.size())' failed.

[0] /Users/graysongall/mambaforge3/envs/moose/libmesh/include/timpi/parallel_implementation.h, line 2665, compiled Aug 21 2024 at 13:55:36
````
 
and the program is hanging here


I changed the vector tag types back from `PARALLEL` to `GHOSTED` and the tests are passing again. If we want to incorporate the change from `PARALLEL` in this PR I think I will need a bit of guidance since I have been struggling to figure out how to fix this issue on my own.

lindsayad commented:
It would be nice to see line numbers for the solid mechanics failure, but you don't have to dig into this much more.

I'm sorry that you ran into all these issues. I take responsibility for that. For what it's worth you've done some good debugging! Ok let's undo the change at https://github.com/idaholab/moose/pull/28177/files#diff-5881bf398171add25bab2869467a34c78f0e54268340077ae34fc7811c0f9422R606. And let's also change your dont-zero tag vectors to be ghosted as well since it sounds like you will be reading from them during residual evaluation

gsgall committed:
Changing tagged vector types back to GHOSTED

- using PARALLEL vector type breaks a few tests in the optimization and solid mechanics modules

gsgall commented:
I have an updated stack trace for the solid mechanics failure. There are a few but they all seem to be the same/similar. Steps to reproduce this, assuming moose is compiled in debug mode, should be

```bash
  cd ~/projects/moose/modules/solid_mechanics/test/tests/generalized_plane_strain 
  mpiexec -n 2 /Users/graysongall/projects/moose/modules/solid_mechanics/solid_mechanics-dbg -i generalized_plane_strain_ref_resid.i --error --error-override --libtorch-device cpu --start-in-debugger="sudo lldb -o 'b MPI_Abort' -o c"
```

Which results in this error output 

```bash 
libMesh terminating:
No index 18 in ghosted vector.
Vector contains [0,12)
And empty ghost array.

[0] ./include/libmesh/petsc_vector.h, line 1086, compiled Aug 15 2024 at 16:57:14

Abort(1) on node 0 (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```

I have also  included the updated stack trace which includes line number information. 

```
  * frame #0: 0x00000001083d34c8 libmpi.12.dylib`MPI_Abort
    frame #1: 0x0000000121c94910 libmesh_dbg.0.dylib`libMesh::libmesh_terminate_handler() + 384
    frame #2: 0x000000018d3640cc libc++abi.dylib`std::__terminate(void (*)()) + 16
    frame #3: 0x000000018d367348 libc++abi.dylib`__cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*) + 88
    frame #4: 0x000000018d36728c libc++abi.dylib`__cxa_throw + 308
    frame #5: 0x0000000122a2fba4 libmesh_dbg.0.dylib`libMesh::PetscVector<double>::map_global_to_local_index(unsigned long long) const + 1560
    frame #6: 0x0000000122a518a0 libmesh_dbg.0.dylib`libMesh::PetscVector<double>::get(std::__1::vector<unsigned long long, std::__1::allocator<unsigned long long>> const&, double*) const + 144
    frame #7: 0x000000011591753c libmoose-dbg.0.dylib`MooseVariableScalar::reinit(this=0x000000014480aa00, reinit_for_derivative_reordering=false) at MooseVariableScalar.C:169:30
    frame #8: 0x0000000115cea3b4 libmoose-dbg.0.dylib`SystemBase::reinitScalars(this=0x0000000143872c18, tid=0, reinit_for_derivative_reordering=false) at SystemBase.C:439:10
    frame #9: 0x0000000114d33e40 libmoose-dbg.0.dylib`FEProblemBase::reinitScalars(this=0x0000000143870400, tid=0, reinit_for_derivative_reordering=false) at FEProblemBase.C:2163:9
    frame #10: 0x0000000115d8bb64 libmoose-dbg.0.dylib`AuxiliarySystem::computeScalarVars(this=0x000000014380be18, type=(_raw_name = "TIMESTEP_END", _name = "TIMESTEP_END", _id = 6)) at AuxiliarySystem.C:636:21
    frame #11: 0x0000000115cec4b8 libmoose-dbg.0.dylib`AuxiliarySystem::compute(this=0x000000014380be18, type=(_raw_name = "TIMESTEP_END", _name = "TIMESTEP_END", _id = 6)) at AuxiliarySystem.C:396:5
    frame #12: 0x0000000114daf6d0 libmoose-dbg.0.dylib`FEProblemBase::computeSystems(this=0x0000000143870400, type=0x0000000116d0b5e8) at FEProblemBase.C:8894:9
    frame #13: 0x0000000114d7cb5c libmoose-dbg.0.dylib`FEProblemBase::execute(this=0x0000000143870400, exec_type=0x0000000116d0b5e8) at FEProblemBase.C:4320:3
    frame #14: 0x0000000113fcd944 libmoose-dbg.0.dylib`FixedPointSolve::solveStep(this=0x0000000104b71310, begin_norm=0x00006000003ae740, end_norm=0x00006000003adf10, transformed_dofs=size=0) at FixedPointSolve.C:470:14
    frame #15: 0x0000000113fccb4c libmoose-dbg.0.dylib`FixedPointSolve::solve(this=0x0000000104b71310) at FixedPointSolve.C:291:28
    frame #16: 0x000000011598d134 libmoose-dbg.0.dylib`TimeStepper::step(this=0x0000000104b96830) at TimeStepper.C:168:52
    frame #17: 0x0000000113fd3c98 libmoose-dbg.0.dylib`Transient::takeStep(this=0x000000011200a800, input_dt=-1) at Transient.C:416:18
    frame #18: 0x0000000113fd2c80 libmoose-dbg.0.dylib`Transient::execute(this=0x000000011200a800) at Transient.C:291:5
    frame #19: 0x00000001162cff10 libmoose-dbg.0.dylib`MooseApp::executeExecutioner(this=0x0000000143866c18) at MooseApp.C:1172:19
    frame #20: 0x00000001162c93e8 libmoose-dbg.0.dylib`MooseApp::run(this=0x0000000143866c18) at MooseApp.C:1554:5
    frame #21: 0x0000000104853658 solid_mechanics-dbg`void Moose::main<SolidMechanicsTestApp>(argc=8, argv=0x000000016b5bd790) at MooseMain.h:47:8
    frame #22: 0x00000001048535cc solid_mechanics-dbg`main(argc=8, argv=0x000000016b5bd790) at main.C:17:3
    frame #23: 0x000000018d023154 dyld`start + 2476 
```

Here is an updated stack trace for the solid mechanics failure

gsgall commented:
> It would be nice to see line numbers for the solid mechanics failure, but you don't have to dig into this much more.
> 
> I'm sorry that you ran into all these issues. I take responsibility for that. For what it's worth you've done some good debugging! Ok let's undo the change at https://github.com/idaholab/moose/pull/28177/files#diff-5881bf398171add25bab2869467a34c78f0e54268340077ae34fc7811c0f9422R606. And let's also change your dont-zero tag vectors to be ghosted as well since it sounds like you will be reading from them during residual evaluation

Thanks @lindsayad, I appreciate it!

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2420354/) on ea38f1f : invalidated by @lindsayad

moosebuild commented:
Job [Python 3.10 (Apptainer)](https://civet.inl.gov/job/2420341/) on ea38f1f : invalidated by @lindsayad

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2420369/) on ea38f1f : invalidated by @lindsayad

lindsayad commented on framework/src/problems/SubProblem.C:
```diff
@@ -135,6 +135,18 @@ SubProblem::vectorTagExists(const TagName & tag_name) const
+}
+
+bool
+SubProblem::vectorTagNotZeroed(const TagID tag)
```
```suggestion
SubProblem::vectorTagNotZeroed(const TagID tag) const
```

lindsayad commented on framework/include/problems/SubProblem.h:
```diff
@@ -144,6 +144,22 @@ class SubProblem : public Problem
+   * @returns false if the tag is not within the set of vectors that are
+   *          intended to not be zero or if the set is empty. returns true otherwise
+   */
+  bool vectorTagNotZeroed(const TagID tag);
```
```suggestion
  bool vectorTagNotZeroed(const TagID tag) const;
```

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2420372/) on ea38f1f wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>bdb543</th>
      <th colspan=4>#28177 ea38f1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.03%</td>
      <td>85.03%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>105213</td>
      <td>105225</td>
      <td>+12</td>
      <td>15</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18527</td>
      <td>18528</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28177/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28177/coverage/framework)


## Modules coverage

Mutiple base shas were found when retrieving head reports:
  bdb543d7e7e11a514a81329d5e5549b502287624
  1926a677783ba66c803fb305330368d0960ff2a7

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/framework)
- `chemical_reactions` (FAILED)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/combined)
- `contact` (FAILED)
- `electromagnetics` (FAILED)
- `external_petsc_solver` (FAILED)
- `fluid_properties` (FAILED)
- `fsi` (FAILED)
- `functional_expansion_tools` (FAILED)
- `geochemistry` (FAILED)
- `heat_transfer` (FAILED)
- `level_set` (FAILED)
- `misc` (FAILED)
- `navier_stokes` (FAILED)
- `optimization` (FAILED)
- `peridynamics` (FAILED)
- `phase_field` (FAILED)
- `porous_flow` (FAILED)
- `ray_tracing` (FAILED)
- `rdg` (FAILED)
- `reactor` (FAILED)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/richards)
- `scalar_transport` (FAILED)
- `solid_mechanics` (FAILED)
- `solid_properties` (FAILED)
- `stochastic_tools` (FAILED)
- `thermal_hydraulics` (FAILED)
- `xfem` (FAILED)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2420372/) on ea38f1f wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 85.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

gsgall commented:
I tried to replicate this failure on my local machine but was unable to after rebasing with next and using the `conda create -n moose-open moose-dev=2024.10.17=openmpi` environment. When running these tests on my machine both are passing.

Were you ever able to take a look at this @loganharbour?

loganharbour commented:
> I tried to replicate this failure on my local machine but was unable to after rebasing with next and using the `conda create -n moose-open moose-dev=2024.10.17=openmpi` environment. When running these tests on my machine both are passing.
> 
> Were you ever able to take a look at this @loganharbour?

I haven't. I'm afraid I've forgotten the context as well. 

Are we looking at https://github.com/idaholab/moose/pull/28177#issuecomment-2305538276?

gsgall commented:
> > I tried to replicate this failure on my local machine but was unable to after rebasing with next and using the `conda create -n moose-open moose-dev=2024.10.17=openmpi` environment. When running these tests on my machine both are passing.
> > Were you ever able to take a look at this @loganharbour?
> 
> I haven't. I'm afraid I've forgotten the context as well.
> 
> Are we looking at [#28177 (comment)](https://github.com/idaholab/moose/pull/28177#issuecomment-2305538276)?

No, [#28177 (comment)](https://github.com/idaholab/moose/pull/28177#issuecomment-2305538276) was a separate issue. @lindsayad had asked if we could change the GHOSTED tagged vectors to PARALLEL and that broke some stuff unrelated to the ray tracing changes. I reverted back to the GHOSTED type and everything was fine except the failing tests for the OpenMPI MOOSE package. 

After that the only thing left to do was to get all of the tests passing and the OpenMPI ones are the only ones with problems. Since neither of us were able to replicate the errors with the OpenMPI Package should I just rebase and push to try testing with the updated packages?

lindsayad commented:
You're referring to the current openmpi failures? I don't believe these are related to your work. @loganharbour I think we need to remove the openmpi target. It just raises false positives, confusing developers. I wonder how many hours @gsgall has wasted thinking about these unrelated failures

loganharbour commented:
> @loganharbour I think we need to remove the openmpi target. It just raises false positives, confusing developers.

I would agree, except all of our built binaries on HPC are all OpenMPI. So I really hesitate at removing them. I've tried to investigate it more, but haven't had any luck. I could probably add in retries for this kind of failure, however.

gsgall commented:
> You're referring to the current openmpi failures? I don't believe these are related to your work. @loganharbour I think we need to remove the openmpi target. It just raises false positives, confusing developers. I wonder how many hours @gsgall has spent thinking about these unrelated failures

I have certainly spent a fair bit of time trying to dig into these or even replicate them on my schools cluster.

gsgall commented:
> > @loganharbour I think we need to remove the openmpi target. It just raises false positives, confusing developers.
> 
> I would agree, except all of our built binaries on HPC are all OpenMPI. So I really hesitate at removing them. I've tried to investigate it more, but haven't had any luck. I could probably add in retries for this kind of failure, however.

Since the issue is unrelated to my additions I will go ahead, rebase with next and push.

lindsayad commented:
> I would agree, except all of our built binaries on HPC are all OpenMPI. So I really hesitate at removing them.

Could we just have it run on next? The people who look at next are a smaller set of folks than those who create PRs, and the former know better what to expect with respect to that openmpi target

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2554158/), step Docs: sync website on ce6c221 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28177/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2554173/), step Generate coverage on ce6c221 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>37b999</th>
      <th colspan=4>#28177 ce6c22</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.13%</td>
      <td>85.13%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>107288</td>
      <td>107301</td>
      <td>+13</td>
      <td>15</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18742</td>
      <td>18742</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28177/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28177/coverage/framework)


## Modules coverage

### Ray tracing

<table>
  <thead>
    <tr>
      <th></th>
      <th>37b999</th>
      <th colspan=4>#28177 ce6c22</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.33%</td>
      <td>95.33%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4364</td>
      <td>4367</td>
      <td>+3</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>214</td>
      <td>214</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28177/coverage_diff/ray_tracing)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28177/coverage/ray_tracing)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28177/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud committed:
Address Alex's review

Add constness to a status-getter

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud reviewed (approved):
I verified that both reviews were addressed.

GiudGiud commented:
Thanks for the contribution @gsgall ! and thanks for investigating the various improvements suggested

gsgall commented:
Happy to help @GiudGiud! And thank you for getting this PR moved along so quickly.
